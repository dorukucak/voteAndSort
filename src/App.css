.outer {
  text-align: center;
  margin-right: 5%;
  margin-left: 5%;
  max-width: 960px;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.header {
  display: flex;
  margin: auto
}

.main {
  display: flex;
  justify-content: center;
  margin: auto
}

.big-button {
  margin: auto;
  height: 100px;
  width: 200px;
  font-size: 20px;

}

.list {
  display: flex;
  flex-direction: column-reverse
}

.add-option {
  display: flex;
  flex-direction: column;
  justify-content: center;
  max-width: 400px;
}

.c{
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  font-size: 12px;
}

.exp{
  display: flex;
}

.actionDiv{
  display: flex;
  justify-content: space-around;
  flex-direction: row;
  width: 250px;
}

