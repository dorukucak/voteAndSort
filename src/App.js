import React from 'react';
import './App.css';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom';
import logo from './hepsiburada-logo.png';

// App creation

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {options: [],
      filters: 'Asc',
      selectedOption: undefined};
  };
  
  // Handles deletion of website entry - Remove button

  handleDeleteOption = (optionToRemove) => {
    this.setState((prevState) => ({
      options: prevState.options.filter((option, index) => optionToRemove !==  index + 1)
    }));
  };
 
  // Handles website addition - Add Link button at /create

  handleAddOption = (option) => {
    if (!option) {
      return 'Enter valid value to add item';
    } else if (this.state.options.indexOf(option) > -1) {
      return 'This option already exists';
    }
    alert(option[0] + ' is added');
    this.setState((prevState) => ({
      options: prevState.options.concat([option])
    }));
  };

  // Up vote handler

   handleUpVote = (siteCount, siteVote) => {     
    this.state.options.map((site, index) => {
      if(index + 1 == siteCount)
      {
      //  const newOptions = this.state.options.slice() //copy the array
        this.state.options[index][2] = siteVote + 1; //execute the manipulations
        
        this.setState((prevState) => 
            ({options: prevState.options.concat([])})) //set the new state
            
    }})    
   };

   // Down vote handler

   handleDownVote = (siteCount, siteVote) => {     
    this.state.options.map((site, index) => {
      if(siteVote == 0) {return}
      else if(index + 1 == siteCount)
      {
      //  const newOptions = this.state.options.slice() //copy the array
        this.state.options[index][2] = siteVote - 1; //execute the manipulations
        
        this.setState((prevState) => 
            ({options: prevState.options.concat([])})) //set the new state
            
    }})    
   };

  // Sets new state of options(website entries) when a new entry is placed

  componentDidMount() {
    try {
      const json = localStorage.getItem('options');
      const options = JSON.parse(json);

      if (options) {
        this.setState(() => ({ options }));
      }
    } catch (e) {
      // Do nothing at all
    }
  }

  // Updates localstorage upon change in state from componentDidMount

  componentDidUpdate(prevProps, prevState) {
  
      const json = JSON.stringify(this.state.options);
      localStorage.setItem('options', json);
    
  }
  componentWillUnmount() {
    console.log('componentWillUnmount');
  };

// Dropdown list that toggles sort on change

onSortChange = (e) => {    
    if (e.target.value === 'Asc') {
    this.setState({
    options: [...this.state.options].sort(function(a, b) {
    if (a[2] < b[2]) {
      return -1;
    }
    if (a[2] < b[2]) {
      return 1;
    }
    return 0;
    })
  });} else if (e.target.value === 'Desc')  {
    this.setState({
    options: [...this.state.options].sort(function(a, b) {
      if (a[2] > b[2]) {
        return -1;
      }
      if (a[2] > b[2]) {
        return 1;
      }
      return 0;
    })
  });}
  };    

 render() { return (
   <div>
      <BrowserRouter>
      <div className='outer'>
          <Header />            
          <Switch>
              <Route path="/" render={() => 
                <Main {...this.state} 
                handleDeleteOption={this.handleDeleteOption}  
                handleUpVote={this.handleUpVote}
                handleDownVote={this.handleDownVote}
                onSortChange={this.onSortChange}                          
                />} exact={true} />
              <Route path="/create"  render={() => <AddLinkPage {...this.state} handleAddOption={this.handleAddOption} />}/>        
          </Switch>
             
          </div> 
   </BrowserRouter>
   </div>
  )};
  
};

// Adding website entries -- appears at /create page

export class AddLinkPage extends React.Component {

  render() {
    return (
      <div>
      <NavLink to="/" activeClassName="is-active">Return to homepage</NavLink>
      <AddOption
      handleAddOption={this.props.handleAddOption}
    />
    </div>
    )
  }
}

// Text inputs for add website -- AddLinkPage child

class AddOption extends React.Component {
    state = {
      error: undefined
    };
    handleAddOption = (e) => {
      e.preventDefault();
      const title = e.target.elements.title.value.trim();
      const link =  e.target.elements.link.value.trim();
      const vote = 0;
      const option = [title, link, vote]
      const error = this.props.handleAddOption(option);
  
      this.setState(() => ({ error }));
  
      if (!error) {
        e.target.elements.title.value = '';
        e.target.elements.link.value = '';
      }
    };
    render() {
      return (
        <div class="c">
          {this.state.error && <p className='add-option-error'>{this.state.error}</p>}
          <form className='add-option' onSubmit={this.handleAddOption}>
            <p class="exp">Link Name:</p>
            <input className='add-option__input' type="text" name="title" placeholder="Add Website Name"/>
            <p class="exp">Link Name:</p>
            <input className='add-option__link' type="text" name="link" placeholder="URL"/>
            <button className='button'>Add Link</button>
          </form>
        </div>
      );
    }
  }

  // Header - logo rendering
  
  const Header = (props) => (
  
    <div className='header'>
      <div className='container'>
      <img src={logo} alt="hepsiburada-logo" width="200" height="125"/>
      </div>      
    </div>
  
  );

// Website entry

const Option = (props) => (
  <div className='option'>
    <p className='option__text'>{props.vote} {props.optionText} {props.link}</p>
    <div className="actionDiv">
    <button className='vote'
      onClick={(e) => {
        props.handleDeleteOption(props.count);
      }}>remove</button>    
      <button className='vote'
      onClick={(e) => {
        props.handleUpVote(props.count, props.vote);
      }}>Up Vote</button>
      <button className='vote'
      onClick={(e) => {
        props.handleDownVote(props.count, props.vote);
      }}>Down Vote</button>
      </div>
  </div>
);

//Websites entry list

export const Options = (props) =>  (
  <div>  
  {props.options.length === 0 && <p className='widget__message'>Please submit a link to get started!</p>}
  <list className='list'>{
    props.options.map((option, index) => (
      <Option 
        key={index}
        optionText={option[0]}
        link={option[1]}
        vote={option[2]}
        count={index + 1}
        handleUpVote={props.handleUpVote}
        handleDownVote={props.handleDownVote}
        handleDeleteOption={props.handleDeleteOption}
      />
    ))
  }</list>
  </div>);

//Main page

export class Main extends React.Component {
 
  render() {
    return (
      <div className='main'>       
        <div className='container'>
          <Link className='big-button' to="/create" activeClassName="is-active">
          <button className='big-button'>SUBMIT A LINK</button>
          </Link>
          <div className='widget'>
            <select
            value={this.props.filters.sortBy}
            onChange={this.props.onSortChange}
            >
            <option value='Asc' >Desc</option>  {/* reverse value due to flex-direction: column-reverse at App.css */}
            <option value='Desc'>Asc</option>
            </select>
            <Options
            options={this.props.options}
            handleUpVote={this.props.handleUpVote}
            handleDownVote={this.props.handleDownVote}
            handleDeleteOptions={this.props.handleDeleteOptions}
            handleDeleteOption={this.props.handleDeleteOption}
            />    
          </div>      
        </div>
      </div>
    );
  }
}


export default App;


