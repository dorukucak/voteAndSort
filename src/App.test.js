import React from 'react';
import { shallow, enzyme } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from './App';
import { render } from '@testing-library/react';
import { Main } from './App';
import { expect, jest } from '@jest/globals';



test('should render App correctly', () => {
  const wrapper = shallow(<App />) 
  expect(toJson(wrapper)).toMatchSnapshot();
});


test('should render Main Page correctly', () => {
  const { getByText } = render(<App />);
  expect(getByText(/Please submit a link to get started!/i)).toBeInTheDocument();
});

test('should show data on Main Page correctly', () => {
  const mockLink = [["hepsiburada", "www.hepsiburada.com", 5], ["n11", "www.n11.com", 2]]
  const wrapper = shallow(<Main options={mockLink} filters={'Asc'}/>);
  expect(toJson(wrapper)).toMatchSnapshot();
})

test('should ascending-sort data on Options correctly', () => {
  const onSortChange = jest.fn();
  const value = 'Desc';
  const mockLink = [["hepsiburada", "www.hepsiburada.com", 5], ["n11", "www.n11.com", 2]];
  const wrapper = shallow(<Main options={mockLink} onSortChange={onSortChange} filters={'Desc'}/>);
  wrapper.find('select').simulate('change', {
    target: { value }
  });
  expect(wrapper.instance('filters')).toBe(value);
})

test('should ascending-sort data on Options correctly', () => {
  const onSortChange = jest.fn();
  const value = 'Asc';
  const mockLink = [["hepsiburada", "www.hepsiburada.com", 5], ["n11", "www.n11.com", 2]];
  const wrapper = shallow(<Main options={mockLink} onSortChange={onSortChange} filters={'Asc'}/>);
  const component = wrapper.instance();
  wrapper.find('select').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('filters')).toBe(value);
})

