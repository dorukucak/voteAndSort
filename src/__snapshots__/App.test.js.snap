// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should ascending-sort data on Options correctly 1`] = `
<div>
  <list
    className="list"
  >
    <Option
      count={1}
      key="0"
      link="www.hepsiburada.com"
      optionText="hepsiburada"
      vote={5}
    />
    <Option
      count={2}
      key="1"
      link="www.n11.com"
      optionText="n11"
      vote={2}
    />
  </list>
</div>
`;

exports[`should render App correctly 1`] = `
<div>
  <BrowserRouter>
    <div
      className="outer"
    >
      <Header />
      <Switch>
        <Route
          exact={true}
          path="/"
          render={[Function]}
        />
        <Route
          path="/create"
          render={[Function]}
        />
      </Switch>
    </div>
  </BrowserRouter>
</div>
`;

exports[`should show data on Main Page correctly 1`] = `
<div
  className="main"
>
  <div
    className="container"
  >
    <Link
      activeClassName="is-active"
      className="big-button"
      to="/create"
    >
      <button
        className="big-button"
      >
        SUBMIT A LINK
      </button>
    </Link>
    <div
      className="widget"
    >
      <select>
        <option
          value="Asc"
        >
          Desc
        </option>
          
        <option
          value="Desc"
        >
          Asc
        </option>
      </select>
      <Options
        options={
          Array [
            Array [
              "hepsiburada",
              "www.hepsiburada.com",
              5,
            ],
            Array [
              "n11",
              "www.n11.com",
              2,
            ],
          ]
        }
      />
    </div>
  </div>
</div>
`;
